src/test/java/org/jabref/logic/layout/format/MarkdownFormatterTest.java

```diff
@@ -21,6 +21,41 @@ void formatWhenFormattingPlainTextThenReturnsTextWrappedInParagraph() {
         assertEquals("<p>Hello World</p>", markdownFormatter.format("Hello World"));
     }
 
+    @Test
+    void formatWhenFormattingHeaderThenReturnsHeaderInHtml() {
```

Comment:
You could convert all of them to ParameterizedTests!

All the tests seems to test some markdown-to-html functionality, which JabRef does not implement by itself, but through a library.

I would remove them and keep `formatWhenFormattingCodeBlockThenReturnsCodeBlockInHtml() {` only.

And then add some tests with empty lines to see how the converter deals with blocks. Are they also removed - or is only the code block the issue?

What about code blocks "recommended" by [MD031 of markdown-lint](https://github.com/DavidAnson/markdownlint/blob/main/doc/md031.md)?

````java
"""
   First line

   Second line

   ```java
   String test;
   ```
"""
````

Reply:
After having written some tests I get what's going on about new line formatting. The current formatter uses HTML `<p>` tags for each line. By default, HTML inserts a new line between each `<p>` tag, which is why in `MarkdownFormatter` we remove the `\n` before returning formatted strings, to avoid jumping two lines instead of one. But in `<pre>` tags which include the code in Mardown, there are no such `<p>` tags to indicate the new lines. When JabRef removes the `\n` before returning the final string, we lose track of those new lines in code blocks.